### PR TITLE
File based implementation of credentials service.

### PR DESCRIPTION
**Pull request by sysexcontrol**
Step 2 of implementation of the credentials API.
Contains the file based service implementation but still without the endpoint and the integration into Hono itself (will be provided in next step).

Signed-off-by: Karsten Frank <Karsten.Frank@bosch-si.com>

### CODE REVIEW

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+    private JsonObject getCredentials(final String tenantId, final String authId, final String type) {
+        Objects.nonNull(tenantId);
+        Objects.nonNull(authId);
+        Objects.nonNull(type);
+
+        final Map<String, JsonArray> credentialsForTenant = credentials.get(tenantId);
+        if (credentialsForTenant != null) {
+            JsonArray authIdCredentials = credentialsForTenant.get(authId);
+            if (authIdCredentials == null)
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: curly braces

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+    @Override
+    public void updateCredentials(String tenantId, JsonObject otherKeys, Handler<AsyncResult<CredentialsResult>> resultHandler) {
+        // TODO: implement
+        CredentialsResult credentialsResult =  CredentialsResult.from(HTTP_NOT_IMPLEMENTED);
+        resultHandler.handle(Future.succeededFuture(credentialsResult));
+    }
+
+    @Override
+    public void removeCredentials(String tenantId, String deviceId, String type, String authId, Handler<AsyncResult<CredentialsResult>> resultHandler) {
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: final params

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+    @Override
+    public void addCredentials(String tenantId, JsonObject otherKeys, Handler<AsyncResult<CredentialsResult>> resultHandler) {
+        // TODO: implement
+        CredentialsResult credentialsResult =  CredentialsResult.from(HTTP_NOT_IMPLEMENTED);
+        resultHandler.handle(Future.succeededFuture(credentialsResult));
+    }
+
+    @Override
+    public void updateCredentials(String tenantId, JsonObject otherKeys, Handler<AsyncResult<CredentialsResult>> resultHandler) {
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: final params

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    }
+
+    @Override
+    public void getCredentials(String tenantId, String type, String authId, Handler<AsyncResult<CredentialsResult>> resultHandler) {
+        CredentialsResult credentialsResult = getCredentialsResult(tenantId, authId, type);
+        resultHandler.handle(Future.succeededFuture(credentialsResult));
+    }
+
+    @Override
+    public void addCredentials(String tenantId, JsonObject otherKeys, Handler<AsyncResult<CredentialsResult>> resultHandler) {
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: final params

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                writeResult.complete();
+            } else {
+                log.warn("could not write credentials to file {}", filename, writeAttempt.cause());
+                writeResult.fail(writeAttempt.cause());
+            }
+        });
+    }
+
+    @Override
+    public void getCredentials(String tenantId, String type, String authId, Handler<AsyncResult<CredentialsResult>> resultHandler) {
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: final params

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+                    }
+                });
+            } else {
+                log.debug("credentials file {} does not exist (yet)", filename);
+            }
+        }
+    }
+
+    @Override
+    protected void doStop(Future<Void> stopFuture) {
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: final params

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    public void setSaveToFile(final boolean enabled) {
+        if (running) {
+            throw new IllegalStateException("credentials registry already started");
+        }
+        this.saveToFile = enabled;
+    }
+
+
+    @Override
+    protected void doStart(Future<Void> startFuture) throws Exception {
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: final params

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+/**
+ * A credentials service that keeps all data in memory but is backed by a file.
+ * <p>
+ * On startup this adapter loads all added credentials from a file. On shutdown all
+ * credentials kept in memory are written to the file.
+ */
+@Repository
+@ConfigurationProperties(prefix = "hono.credentials")
+@Profile({"default", "credentials-file"})
+public class FileBasedCredentialsService extends BaseCredentialsService {
```
Comment (sophokles73) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: either make class final or mark operations not to be overridden final

File: server/src/test/resources/credentials.json [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+      "enabled" : false,
+      "secrets" : [ {
+        "hash-function" : "sha-512",
+        "salt": "aG9ubw==",
+        "pwd-hash" : "C9/T62m1tT4ZxxqyIiyN9fvoEqmL0qnM4/+M+GHHDzr0QzzkAUdGYyJBfxRSe4upDzb6TSC4k5cpZG17p4QCvA=="
+      }]
+    }
+
+  ]
+} ]
```
Comment (sophokles73) [inline server/src/test/resources/credentials.json]: missing EOL

File: service-base/src/main/java/org/eclipse/hono/service/credentials/BaseCredentialsService.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            switch (subject) {
+            case OPERATION_GET: {
+                String type = payload.getString(FIELD_TYPE);
+                String authId = payload.getString(FIELD_AUTH_ID);
+                log.debug("getting credentials [{}:{}] of tenant [{}]", type, authId, tenantId);
+                getCredentials(tenantId, type, authId, result -> reply(regMsg, result));
+            }
+                break;
+            default:
+                log.info("subject [{}] not supported", subject);
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/credentials/BaseCredentialsService.java]: *operation* or *action* not supported

File: service-base/src/main/java/org/eclipse/hono/service/credentials/BaseCredentialsService.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+     * to be returned to clients as the result of a registration operation.
+     * TODO: params
+
+     * @return The JSON structure.
+     */
+    protected final static JsonObject getResultPayload(final String deviceId, final  String type, final String authId, final boolean enabled, final JsonArray secrets) {
+        return new JsonObject().
+                put(FIELD_DEVICE_ID, deviceId).
+                put(FIELD_TYPE, type).
+                put(FIELD_AUTH_ID,authId).
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/credentials/BaseCredentialsService.java]: missing space after comma (next two lines as well)

File: server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java (hunk 8 above)
Comment (sysexcontrol) [inline server/src/main/java/org/eclipse/hono/service/credentials/impl/FileBasedCredentialsService.java]: Decided to mark the single operations for more flexibility. The currently still "not implemented" credentials API methods will remain not final but will be changed to final when implementation was added.

**Comment by sysexcontrol** [comment]
Fixed all the commented points, can you please have a look again? Thx

**Pull request merged**